New chat input & suggestions area; new handlers to work with the improved logic (#880)

alwx opened the pull request:
This PR seems huge (and it is huge), but because new design of input and suggestions area is not only a graphical change but also a big change in the logic, this PR is important.

The new approach introduces several new UI elements and completely new input and suggestions area. The detailed description will be added here when the PR will be ready (most probably on Friday or Saturday).

alwx commented:
This PR completely changes the way suggestions and command handlers work. It both simplify things and adds several cool and amazing features for DApp developers. The main idea behind these changes is that according to the new design we now need to have an ability to specify several parameters in one field and also not only suggest something but "guide" a user step-by-step. Unfortunately, the previous schema of input field organization was a bit hacky (for example, there was no way to determine if a text is a command or not dynamically, so we needed to add "c " prefix for each command) and hardly supported multiple arguments in one field.
The new input field offers multiple parameters, cool suggestions for each parameter (by the way, they are no longer called "suggestions"), supports placeholder and secure-text-entry for passwords. And it uses one multiline input field for everything! Sounds like magic.

Unfortunately, it was not possible to make all these big changes without changing DApp js files.

### 1. No more "c " prefixes and hacks
The new input uses the new set of handlers to control everything. The coolest thing here is that all handlers for inputs and suggestions are placed in `status-im.chat.handlers.input`, and it's just about 200 lines of code!

### 2. We don't need to use `suggestions` for everything anymore

Previously, our DApp JS files looked this way:
```
function wallet(params, context) {
    var url = 'https://status.im/dapps/wallet';

    if (context.debug) {
        url = 'http://127.0.0.1:3450';
    }

    if (params.url && params.url !== "undefined" && params.url != "") {
        url = params.url;
        if (!/^[a-zA-Z-_]+:/.test(url)) {
            url = 'http://' + url;
        }
    }

    return {webViewUrl: url};
}

status.command({
    name: "browse",
    title: I18n.t('browse_title'),
    description: I18n.t('browse_description'),
    color: "#ffa500",
    fullscreen: true,
    suggestionsTrigger: 'on-send',
    params: [{
        name: "url",
        suggestions: wallet,
        type: status.types.TEXT
    }]
});
```

In order to call a web browser, we needed to use `suggestions` function (wat?) and `suggestionsTrigger`. Here is the new `browse` command:

```
status.command({
  name: "browse",
  fullscreen: true,
  title: 'Browser',
  params: [{
    name: "url",
    placeholder: "url",
    suggestions: browseSuggestions,
    type: status.types.TEXT
  }],
  onSend: function (params, context) {
    var url = params.args.url;
    if (!/^[a-zA-Z-_]+:/.test(url)) {
      url = 'http://' + url;
    }

    return {
      title: "Browser",
      dynamicTitle: true,
      markup: status.components.bridgedWebView(url)
    };
  }
});
```

As you may see, it uses `onSend` function (which is, of course, optional and simply overrides the default behavior). This fact, first of all, means that we now able to use BOTH `suggestions` for parameters and `onSend` functions to display some content. 

### 3. `markup`
You may also notice from the previous example that `browse` command returns not only a `webViewUrl` (this is how it worked before) but a Status component (webView) in `markup` field instead. The same is true for any other function that returns something to display. This fact means that `onSend` function may display any other status component (for example, some native text) and suggestions can now also use a fully-functional web browser, and this will definitely be useful for implementing new complex functions like this one:

<img width="280" alt="screen shot 2017-03-20 at 15 52 33" src="https://cloud.githubusercontent.com/assets/911127/24100480/7e717574-0d85-11e7-86cd-7a7092d93dd8.png">

The coolest thing is that we may allow developers to return not only a `markup`, but something else. In this example, the `browse` command returns a default title for the web browser (`title` property) and also tells Status to dynamically change the title (`dynamicTitle` property). 
Just imagine: it, for instance, gives us an ability to let DApp developers to specify a set of additional buttons for their DApps and makes possible to implement something that looks this way:

<img width="281" alt="screen shot 2017-03-20 at 15 58 18" src="https://cloud.githubusercontent.com/assets/911127/24100805/d03e0cae-0d86-11e7-8634-87b10284648a.png">

DApps can now be better integrated with Status and may also "communicate" with Status.

## Limitations and things to be implemented

1. Animations;
2. We have no obvious ability to support `keyboard-type` when our input contains mixed content;
3. `suggestions` and `on-send` parameters for new functions like `/send` or `/request`. Everything works well without these parameters, but if we want to have nice-looking "suggestions", we should implement them too (btw, I call them `parameter-boxes` in code — it is not completely right to call them suggestions because they are able to do much more than simply "suggest");
4. JS suggestions in `console` chat.

alwx commented:
It's finished and can be reviewed.
@rasom, @jarradh, @flexsurfer, @GustavoNunes you're welcome :)

## To be implemented
1. Animations (will be done separately in `feature/chat-ui-animations` branch);
2. We have no obvious ability to support `keyboard-type` when our input contains mixed content (for instance, a command string may contain something like `/send 0.5 ETH "Main wallet" 0x0434f2...`), but I believe that there is a way of solving this issue. I was experimenting with it in `experiment/password-input` branch.
3. `suggestions` and `on-send` parameters for new functions like `/send` or `/request`. Everything works well without these parameters, but if we want to have nice-looking "suggestions", we should implement them too. There is a special PR to solve this issue: https://github.com/status-im/status-react/pull/939

rasom commented:
1. Dapp's url isn't opened
  ![dapps_url](https://cloud.githubusercontent.com/assets/2364994/24397056/dabf9c38-13ad-11e7-865d-897c2fde5658.gif)

2. /browse can't be opened on fullscreen, is this because animations aren't implemented? 
 ![fullscreen](https://cloud.githubusercontent.com/assets/2364994/24397106/fa638a5e-13ad-11e7-91f4-ea2ffba4cc77.gif)

3. Is this big blue horizontal scrollbar by design?
![scrollbar](https://cloud.githubusercontent.com/assets/2364994/24397176/249d1c72-13ae-11e7-9990-b09750ac8e76.gif)

flexsurfer commented on ios/StatusIm/Info.plist:
```diff
@@ -58,6 +58,7 @@
 	<string>We need to access your photo storage to give you an ability to select photos</string>
 	<key>UIAppFonts</key>
 	<array>
+		<string>RobotoMono-Medium.ttf</string>
```
why do we want to use Roboto on iOS, i know maybe it's a question to @andmironov

alwx replied:
yes, definitely. iOS designs also use Roboto

flexsurfer commented on src/status_im/chat/handlers.cljs:
```diff
@@ -42,17 +42,21 @@
-    (assoc db :show-info show-info)))
+  (fn [{:keys [current-chat-id chat-ui-props] :as db} [_ ui-element chat-id]]
+    (let [chat-id (or chat-id current-chat-id)]
+      (assoc-in db
```
`(update-in db [:chat-ui-props chat-id ui-element] not)` ?

flexsurfer commented on src/status_im/chat/styles/input/emoji.cljs:
```diff
@@ -0,0 +1,16 @@
+(ns status-im.chat.styles.input.emoji
+  (:require [status-im.components.styles :as common]))
```
usually we use `:refer` , we need some style convention for this

alwx replied:
it depends — sometimes we also use `:as`. 
I think that it's bad to use `:refer` and we should use `:as` instead *everywhere*, so I decided to use it here as an example.

rasom replied:
`:as` is definitely more explicit than `:refer`, so it's better. About using `:as` everywhere, this is more a matter of taste.

rasom commented:
**[ANDROID]**

![screen shot 2017-03-28 at 11 41 49](https://cloud.githubusercontent.com/assets/2364994/24397839/198fc5b2-13b0-11e7-95ee-99fbc4c678a2.png)

![screen shot 2017-03-28 at 12 04 34](https://cloud.githubusercontent.com/assets/2364994/24397888/3e88f7a8-13b0-11e7-87bf-187490ab6eda.png)

1. Wrong size and color
2. Button is missed 
3. Wrong font
4. Command's name isn't centered horizontally 
5. Line separators are absent in the design
6. Wrong radius
7. Placeholder is missed

flexsurfer commented on src/status_im/chat/styles/input/input.cljs:
```diff
@@ -0,0 +1,119 @@
+          :position            :absolute
+          :text-align-vertical :center
+          :height              min-input-height}
+         (if platform/android?
```
now we have `defnstyle` macro

jarradh commented:
commands should be scrollable but no visible scroll bar

rasom commented:
**[iOS]**

![screen shot 2017-03-28 at 12 56 25](https://cloud.githubusercontent.com/assets/2364994/24399810/ab757976-13b6-11e7-86d0-f272f2b337b0.png)


![img_1511](https://cloud.githubusercontent.com/assets/2364994/24400208/41769bb6-13b8-11e7-9669-26e5b1dcb766.PNG)

1. Extra element
2. Distance is too small
3. Container's height is smaller than in design 
4. Wrong icon
5. Different styles
6. Icon looks differently 
7. Wrong position of the placeholder

rasom commented on src/status_im/chat/models/input.cljs:
```diff
@@ -0,0 +1,117 @@
+(defn current-chat-argument-position
+  [command input-text]
+  (if command
+    (let [current (-> (:args command) (count))]
```
better to use just `(count (:args command))`, or destruct `command` in parameters declaration and use `(count args)`, or at least change it to `(-> (:args command) count)`

rasom commented on src/status_im/chat/models/input.cljs:
```diff
@@ -0,0 +1,117 @@
+(defn join-command-args [args]
+  (->> args
+       (map (fn [arg]
+              (if (= (.indexOf arg const/spacing-char) -1)
```
`clojure.string/index-of` or `clojure.string/inculdes?`

rasom commented on src/status_im/chat/models/input.cljs:
```diff
@@ -0,0 +1,117 @@
+                    [new-list command-continues?]))
+                [[] false]
+                splitted)
+        (first))))
```
In general, no need to use threading macro when it contains less than 2 forms

rasom commented on resources/commands.js:
```diff
@@ -694,79 +694,75 @@ status.command({
-    }]
+    }],
+    onSend: function (params, context) {
+        var url = params.args.url;
```
What is the reason to use `args` here? What also will be included into `params`?

alwx replied:
`params` also contains a `metadata` object

see `:set-chat-input-metadata` handler and `::request-command-data` handler for more details

rasom replied:
I looked at `:set-chat-input-metadata` and `::request-command-data` and both of them don't make clear what this `metadata` contains and how it can be useful in commands handlers inside the jail.

alwx commented:
I've updated the code and fixed the issues you've mentioned.

However, there is the inconsistency between how suggestions area looks on Android and iOS in design sketches. I think it's a mistake. 
@andmironov ?

<img width="600" alt="inconsit" src="https://cloud.githubusercontent.com/assets/911127/24503211/bb8fb5e8-1559-11e7-8729-17bcbb6253f2.png">

andmironov commented:
yep, it was a mistake, thanks @alwx I updated the Android_Chat board https://zpl.io/2lkHvv

jarradh commented:
can we get this in today?

alwx commented:
@jarradh it's already implemented

rasom commented:
@jarradh, probably we can. 

I have a question regarding password confirmation, it was removed in this PR. Do we really need to do this?

rasom commented:
Can't send transaction, seems that address is not passed in the `:context` 

<img width="380" alt="screen shot 2017-03-31 at 11 34 03" src="https://cloud.githubusercontent.com/assets/2364994/24542986/69c52150-1606-11e7-88dd-96ceeecdc30c.png">

rasom commented:
Placeholder on iOS real device
![img_1518](https://cloud.githubusercontent.com/assets/2364994/24544219/200e2944-160b-11e7-8d11-a5a9889b60ae.PNG)

alwx commented:
@rasom am I right that placeholder looks fine on simulator?

rasom commented:
Yes, it looks fine there

<img width="379" alt="screen shot 2017-03-31 at 12 28 28" src="https://cloud.githubusercontent.com/assets/2364994/24544896/bfe8a1cc-160d-11e7-8568-7892da6c0c62.png">

rasom commented:
Dapps are not opened fullscreen on iOS real device

![sa](https://cloud.githubusercontent.com/assets/2364994/24545370/8fbbe0ca-160f-11e7-9b9f-bef33198a221.gif)

alwx commented:
Done.

flexsurfer commented:
ios

1. By clicking on the command, the keyboard disappears, although you expect the command to be entered
2. on /send command nothing happens
3. Extra borders 
![image](https://cloud.githubusercontent.com/assets/11790366/24656221/4587ac08-194a-11e7-9a33-fc773cb26e64.png)
![image](https://cloud.githubusercontent.com/assets/11790366/24656617/e35247e4-194b-11e7-82ce-929595dbb3c6.png)
4. keyboard closed on submit text  (imo it shouldn't)
5. different command titles (i thought mockups should have real texts)
![image](https://cloud.githubusercontent.com/assets/11790366/24656316/b08650fe-194a-11e7-919c-747589904237.png)
![image](https://cloud.githubusercontent.com/assets/11790366/24656388/03e6e088-194b-11e7-8784-020bce0badb7.png)
6. after deleting command placeholder is wrong aligned
![image](https://cloud.githubusercontent.com/assets/11790366/24657742/2aac9ef6-1950-11e7-92d0-97855332efb2.png)
7. unnecessary commands in wallet
![image](https://cloud.githubusercontent.com/assets/11790366/24658150/865506de-1951-11e7-9fa2-f14e60fedef8.png)
8. Browse URL suggestions not implemented?
![image](https://cloud.githubusercontent.com/assets/11790366/24656478/69ffa3d2-194b-11e7-9d14-d21322d09286.png)
9. browser looks different (not implemented ?)
![image](https://cloud.githubusercontent.com/assets/11790366/24656690/1a64006a-194c-11e7-8138-3c70c616c6d4.png)
10. Animations for closing panels are not implemented
11. Extra separator
![image](https://cloud.githubusercontent.com/assets/11790366/24658389/3c4ebcc8-1952-11e7-9a53-72e56acfc30c.png)
12. impossible to enter emoji  
![image](https://cloud.githubusercontent.com/assets/11790366/24658440/5fd018e0-1952-11e7-825a-1c7b09473bf2.png)
13. unnecessary commands in group chat
![image](https://cloud.githubusercontent.com/assets/11790366/24658603/d567b4c8-1952-11e7-90e6-2b459c36a338.png)
14. no suggestions for command
![image](https://cloud.githubusercontent.com/assets/11790366/24658728/3bd4a0c2-1953-11e7-973c-f3a63a5173a0.png)

flexsurfer commented:
impossible to enter password
![image](https://cloud.githubusercontent.com/assets/11790366/24660208/9c13f330-1957-11e7-96e4-c4399617df6e.png)

flexsurfer commented:
Android
1. password placeholder works weird
![image](https://cloud.githubusercontent.com/assets/11790366/24661637/92c4e826-195b-11e7-9188-06e7ebe66510.png)
2. when I'm typing password placeholder is superimposed on the text entered for a short period of time
3. input field doesn't scroll when entered more than 3 lines 
![image](https://cloud.githubusercontent.com/assets/11790366/24661932/63dcd9be-195c-11e7-817c-bcaa55e2951f.png)
4. unnecessary separators
![image](https://cloud.githubusercontent.com/assets/11790366/24662051/bae847fc-195c-11e7-9223-f7772404e954.png)

flexsurfer commented:
when i was working with the input field i noticed that it require a lot of CPU when writing text, and sometimes it works with glitches on Android. I found that each inputted symbol generates at least 15 events, and if you're writing fast it's a large flow of events and calculations, maybe we need some optimization here because chatting fast using simple text is the main purpose of chats

rasom commented:
@flexsurfer yes we need this, especially when `text-change` inits a lot of computations, but not in this PR.

alwx commented:
@flexsurfer thank you for comments. I will fix them today.

jarradh commented:
what are the events generated?

alwx commented:
iOS:

1. Done
2. Done. You should update your Status iOS module to make it work.
3. Done. However, some of the borders you're talking about are not related to this task.
4. Done
5. Done
6. Done
7. Done
8. *Not implemented*. Will be implemented separately.
9. Done
10. Done
11. *Not done*. Not related to this task.
12. Don't see any problem here. 
13. Done

Android:

1. Done
2. Done
3. *Not done*. As I see, it works the same way in `develop`, and probably we need some time to find a reason.
4. Done

flexsurfer commented:
![image](https://cloud.githubusercontent.com/assets/11790366/24749634/7a7c4f3a-1acc-11e7-904e-6936c7dbb9b3.png)
i don't know how to reproduce this , but it happens sometimes

alwx commented:
@flexsurfer I will create a separate branch for it

alwx commented:
This issue is fixed in #982
